core/aggregated_bloom_filter.go

```diff
@@ -0,0 +1,288 @@
+// When adding a key for a particular block, the indices mapped by Bloom hash functions are determined,
+// and the bits at those row-column intersections are set for that block.
+//
+// Visually, this can be thought of as "rotating" the per-block Bloom filters into columns of a matrix.
+//
+// -----| Block 0 | Block 1 | Block 2 | ... | Block 9 |
+// Idx0 |   0     |    0    |    0    | ... |   0     |
+// Idx1 |   1     |    0    |    1    | ... |   0     |
+// Idx2 |   0     |    1    |    0    | ... |   0     |
+// Idx3 |   1     |    0    |    0    | ... |   0     |
+// Idx4 |   1     |    0    |    1    | ... |   1     |
+// Idx5 |   0     |    0    |    0    | ... |   0     |
+// Idx6 |   0     |    0    |    1    | ... |   0     |
+// Idx7 |   0     |    1    |    0    | ... |   0     |
+//
+//	<-----------Bloom Filter Indices----------->
+//
+// To query for a key, the AggregatedBloomFilter:
+//
+//  1. Determines the relevant indices for the key using the same hash functions.
+//  2. Performs a bitwise AND over the selected rows, producing a bit vector.
+//  3. The set bits in this result indicate block numbers within the filter's range
+//     where the key may be present (with the usual caveat of possible false positives).
+//     Note: The set bit positions are *relative to the filter's range start* (i.e., the
+//     range's first block number), not absolute global block numbers.
+//
+// Query example for a key mapping to indices Idx1 and Idx4:
+//
+// Select rows 1 and 4 (Idx1 & Idx4):
+// Idx1:    1    0    1   ...   0
+// Idx4:    1    0    1   ...   1
+//
+//		       --------------------------
+//		AND:    1    0    1   ...   0
+//
+//	After AND: Resulting vector is 1 0 1 ... 0
+//
+//	This means Block 0 and Block 2 are possible matches for this key. Block numbers are
+//
+// This approach allows for efficient, bulk event queries on blockchain data
+// without needing to individually examine every single block’s Bloom filter.
+//
+// Using this method, you can quickly identify candidate blocks for a key, improving
+// the performance of large-range event queries.
+type AggregatedBloomFilter struct {
+	bitmap    []bitset.BitSet
+	fromBlock uint64
+	toBlock   uint64
+}
+
+const (
+	AggregateBloomBlockRangeLen uint64 = 8192
+)
+
+var (
+	ErrAggregatedBloomFilterBlockOutOfRange error = errors.New("block number is not within range")
+	ErrBloomFilterSizeMismatch              error = errors.New("bloom filter len mismatch")
+	ErrMatchesBufferNil                     error = errors.New("matches buffer must not be nil")
+	ErrMatchesBufferSizeMismatch            error = errors.New("matches buffer size mismatch")
+)
+
+// NewAggregatedFilter creates a new AggregatedBloomFilter starting from the specified block number.
+// It initialises the bitmap array with empty bitsets of size AggregateBloomBlockRangeLen.
+func NewAggregatedFilter(fromBlock uint64) *AggregatedBloomFilter {
```

Comment:
I think we can return by value here. At it's core it is a reference type (it holds a pointer and metadata) and only three words in size